clients/src/main/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManager.java

```diff
@@ -0,0 +1,347 @@
+            multiNodeRequest.partitionsToRetry.addAll(partialResult.partitionsToRetry);
+
+            if (multiNodeRequest.expectedResponses.decrementAndGet() == 0) {
+                ListOffsetResult result =
+                        new ListOffsetResult(multiNodeRequest.fetchedTimestampOffsets,
+                                multiNodeRequest.partitionsToRetry);
+                multiNodeRequest.resultFuture.complete(result);
+            }
+        } catch (RuntimeException e) {
+            multiNodeRequest.resultFuture.completeExceptionally(e);
+        }
+    }
+
+    private static class ListOffsetsRequestState {
+
+        private final Map<TopicPartition, ListOffsetData> fetchedOffsets;
+        private final Map<TopicPartition, Long> remainingToSearch;
+        private final List<MultiNodeRequest> multiNodeRequests;
+        private final CompletableFuture<Map<TopicPartition, Long>> globalResult;
+        final boolean requireTimestamps;
+        final OffsetFetcherUtils offsetFetcherUtils;
+        final IsolationLevel isolationLevel;
+
+        private ListOffsetsRequestState(Map<TopicPartition, Long> timestampsToSearch,
+                                        boolean requireTimestamps,
+                                        OffsetFetcherUtils offsetFetcherUtils,
+                                        IsolationLevel isolationLevel) {
+            remainingToSearch = Collections.synchronizedMap(timestampsToSearch);
+            fetchedOffsets = Collections.synchronizedMap(new HashMap<>());
+            multiNodeRequests = Collections.synchronizedList(new ArrayList<>());
+            globalResult = new CompletableFuture<>();
+            this.requireTimestamps = requireTimestamps;
+            this.offsetFetcherUtils = offsetFetcherUtils;
+            this.isolationLevel = isolationLevel;
+        }
+
+        void addMultiNodeRequest(MultiNodeRequest multiNodeRequest) {
+            multiNodeRequest.resultFuture.whenComplete((multiNodeResult, error) -> {
+                // Done sending request to a set of known leaders
+                multiNodeRequests.remove(multiNodeRequest);
+                if (error == null) {
+                    fetchedOffsets.putAll(multiNodeResult.fetchedOffsets);
+                    remainingToSearch.keySet().retainAll(multiNodeResult.partitionsToRetry);
+                    offsetFetcherUtils.updateSubscriptionState(
+                            multiNodeResult.fetchedOffsets,
+                            isolationLevel);
+
+                    if (remainingToSearch.size() == 0) {
+                        ListOffsetResult listOffsetResult = new ListOffsetResult(fetchedOffsets,
+                                remainingToSearch.keySet());
+                        globalResult.complete(listOffsetResult.offsetAndMetadataMap());
+                    }
+                } else {
+                    // TODO: check if retriable errors should be considered/retried here
```

Comment:
it seems like the original code just add the retriable partitions to the results?

Reply:
i see
```
 case NOT_LEADER_OR_FOLLOWER:
                    case REPLICA_NOT_AVAILABLE:
                    case KAFKA_STORAGE_ERROR:
                    case OFFSET_NOT_AVAILABLE:
                    case LEADER_NOT_AVAILABLE:
                    case FENCED_LEADER_EPOCH:
                    case UNKNOWN_LEADER_EPOCH:
                        log.debug("Attempt to fetch offsets for partition {} failed due to {}, retrying.",
                                topicPartition, error);
                        partitionsToRetry.add(topicPartition);
```

Reply:
It does add them but only to the partial results, to register them as remainingToSearch. We do similarly. This comment was mainly referring to the errors like the disconnected leader that the original code was checking upfront and we decided to maybe let fail on the poll. After yesterday's conversation regarding the NodeStatusDetector @kirktrue introduced, I can introduce that